### PR TITLE
Sync some changes from KvikIO's `Array`

### PR DESCRIPTION
**Pull request by jakirkham**
Pulls in a few changes from KvikIO's `Array`. Namely...

* `asarray` for easy conversion to `Array` (no-op if already an `Array`)
* Include types for a few more properties

### CODE REVIEW

File: python/ucxx/ucxx/_lib/arr.pyx [hunk 1]
```diff
@@ -295,3 +297,10 @@ cdef inline Py_ssize_t _nbytes(Py_ssize_t itemsize,
     for i in range(ndim):
         nbytes *= shape_mv[i]
     return nbytes
+
+
+cpdef Array asarray(obj):
```
Comment (madsbk) [inline python/ucxx/ucxx/_lib/arr.pyx:302]: I think it would be good with a docstring

**Review by madsbk — COMMENTED** [review]
Looks good, I only have one suggestion

File: python/ucxx/ucxx/_lib/arr.pyx (hunk 1 above)
Comment (pentschev) [reply python/ucxx/ucxx/_lib/arr.pyx:302]: +1

Comment (jakirkham) [reply python/ucxx/ucxx/_lib/arr.pyx:302]: Copying over KvikIO's recently added docstring ( https://github.com/rapidsai/kvikio/pull/475#pullrequestreview-2338640849 )

```suggestion
cpdef Array asarray(obj):
    """Coerce other objects to ``Array``. No-op for existing ``Array``s.

    Args:
        obj: Object exposing the Python buffer protocol or ``__cuda_array_interface__``

    Returns:
        Array: An instance of the ``Array`` class
    """
```

File: python/ucxx/ucxx/_lib/arr.pyi [hunk 2]
```diff
@@ -1,10 +1,12 @@
...
 # SPDX-License-Identifier: BSD-3-Clause
 
-from typing import Tuple
+from typing import Generic, Tuple, TypeVar
 
-class Array:
-    def __init__(self, obj: object): ...
+T = TypeVar("T")
+
+class Array(Generic[T]):
```
Comment (wence-) [inline python/ucxx/ucxx/_lib/arr.pyi:8]: Same comment here as in ucx-py

File: python/ucxx/ucxx/_lib/arr.pyx [hunk 3]
```diff
@@ -295,3 +297,18 @@ cdef inline Py_ssize_t _nbytes(Py_ssize_t itemsize,
...
+    """Coerce other objects to ``Array``. No-op for existing ``Array``s.
+
+    Args:
+        obj: Object exposing the Python buffer protocol or ``__cuda_array_interface__``
+
+    Returns:
+        Array: An instance of the ``Array`` class
```
Comment (pentschev) [inline python/ucxx/ucxx/_lib/arr.pyx:309]:
```suggestion
    Parameters
    ----------
    obj: object
        Object exposing the Python buffer protocol or ``__cuda_array_interface__``.

    Returns
    -------
    array: Array
        An instance of the ``Array`` class.
```

Comment (pentschev) [reply python/ucxx/ucxx/_lib/arr.pyx:309]: Let's also attempt to keep the docstrings formatting used by the rest of the project.

**Review by pentschev — CHANGES_REQUESTED** [review]

File: python/ucxx/ucxx/_lib/arr.pyi (hunk 2 above)
Comment (jakirkham) [reply python/ucxx/ucxx/_lib/arr.pyi:8]: Same answer: https://github.com/rapidsai/ucx-py/pull/1080#discussion_r1783965174

**Review by pentschev — APPROVED** [review]
LGTM here too, similar to https://github.com/rapidsai/ucx-py/pull/1080#pullrequestreview-2345011269, it's probably good to have @wence- voice his opinion one more time. Thanks @jakirkham .

**Review by wence- — APPROVED** [review]

**Comment by jakirkham** [comment]
Thanks all! 🙏

**Pull request merged**